src/ControlSystem/Measurements/BNSCenterOfMass.hpp

```diff
@@ -0,0 +1,201 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <string>
+
+#include "ControlSystem/Component.hpp"
+#include "ControlSystem/Protocols/Measurement.hpp"
+#include "ControlSystem/Protocols/Submeasurement.hpp"
+#include "ControlSystem/RunCallbacks.hpp"
+#include "DataStructures/DataBox/Tag.hpp"
+#include "DataStructures/LinkedMessageId.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/Structure/ObjectLabel.hpp"
+#include "NumericalAlgorithms/LinearOperators/DefiniteIntegral.hpp"
+#include "Parallel/Reduction.hpp"
+#include "Time/Tags.hpp"
+#include "Utilities/ProtocolHelpers.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+class DataVector;
+template <size_t VolumeDim>
+class ElementId;
+template <size_t Dim>
+class Mesh;
+namespace Parallel {
+template <typename Metavariables>
+class GlobalCache;
+}  // namespace Parallel
+namespace domain::Tags {
+template <size_t Dim>
+struct Mesh;
+template <size_t Dim, typename Frame>
+struct Coordinates;
+}  // namespace domain::Tags
+namespace Events::Tags {
+template <typename InputFrame, typename OutputFrame>
+struct ObserverDetInvJacobian;
+}
+namespace grmhd::ValenciaDivClean {
+struct TildeD;
+}
+/// \endcond
+
+namespace control_system::measurements {
+
+template <typename ControlSystems>
+struct PostReductionSendBNSStarCentersToControlSystem;
+
+namespace Tags {
+/// \ingroup DataBoxTagsGroup
+/// \ingroup ControlSystemGroup
+/// DataBox tag for location of neutron star center
+/// (or more accurately, center of mass of the matter
+/// in the x>0 (label A) or x<0 (label B) region, in grid
+/// coordinates.
+template <::domain::ObjectLabel Center>
+struct NeutronStarCenter : db::SimpleTag {
+  using type = std::array<double, 3>;
+};
+}  // namespace Tags
+
+/// Factored Center of Mass calculation (for easier testing)
+/// This function computes the integral of tildeD (assumed to be the
+/// conservative baryon density in the inertial frame), as well as
+/// its first moment in the grid frame. The integrals are limited to
+/// \f$x>0\f$ (label A) or \f$x<0\f$ (label B).
+/// inv_det_jacobian is the inverse determinant of the jacobian of the
+///                  map between logical and inertial coordinates
+/// xGrid contains the coordinates in the grid frame
+static inline void center_of_mass_integral_on_element(
+    const gsl::not_null<double*> MassA, const gsl::not_null<double*> MassB,
+    const gsl::not_null<std::array<double, 3>*> FirstMomentA,
+    const gsl::not_null<std::array<double, 3>*> FirstMomentB,
+    const Mesh<3>& mesh, const Scalar<DataVector>& inv_det_jacobian,
+    const Scalar<DataVector>& tildeD,
+    const tnsr::I<DataVector, 3, Frame::Grid> xGrid) {
+  // Get Jacobian and mask for left/right stars
+  const DataVector det_jacobian = 1. / get(inv_det_jacobian);
+  const DataVector PositiveX =
+      (xGrid.get(0) + abs(xGrid.get(0))) / (2.0 * abs(xGrid.get(0)) + 1.e-16);
```

Comment:
I believe `max(get<0>(xGrid), 0.0)` works.  (And prefer `get<N>(...)` to `.get(N)` if `N` is a constant.)